### PR TITLE
DrivAerML: attention temperature annealing (new architectural feature)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Attention temperature annealing is a well-established technique from the attention literature: introduce a scheduled temperature parameter τ that scales attention logits by `1/τ`. Starting with τ>1 (softer, broader attention) and annealing toward τ=1 (standard sharp softmax) over training helps the model explore more broadly early on and converge to focused attention patterns later.

This was trialled on the noam branch and showed approximately **-11% improvement** on DrivAerML. We now want to confirm whether it also works on the **radford recipe** (different optimizer/LR schedule family) and validate it as a standalone architectural improvement.

This is a **new architectural feature** — not a hyperparameter tweak of the existing recipe. Each attention head gains an explicit temperature schedule rather than relying purely on softmax sharpness throughout training.

Reference: https://arxiv.org/abs/2302.01107

## Instructions

**Step 1: Read the noam implementation (reference)**

```bash
git diff radford...origin/noam -- target/icml2026/train.py | grep -A 20 -B 5 "temperature\|temp_anneal\|attn_temp"
```

Use this as a guide. If the noam implementation is clean and portable, port it directly. If it is entangled with noam-specific code, use the fallback below.

**Step 2: Implement in `target/icml2026/train.py`**

Add these CLI flags:
- `--attn-temp-annealing` (store_true, default False): enable feature
- `--attn-temp-start` (float, default 2.0): initial τ at epoch 0
- `--attn-temp-end` (float, default 1.0): final τ at last epoch

Annealing schedule at epoch `e` of `E` total:
```python
tau = attn_temp_start - (attn_temp_start - attn_temp_end) * (e / E)
```

Scale attention logits by `1/tau` before softmax in every attention layer. Pass `tau` via a module attribute updated at the start of each epoch.

**Fallback simple implementation:**
```python
# In attention forward pass, before softmax:
logits = logits / tau  # tau updated each epoch by trainer
```

**Step 3: Smoke test (3 epochs)**

```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=3 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --attn-temp-annealing --attn-temp-start 2.0 --attn-temp-end 1.0 \
  --wandb_group dm-innovation-track --name smoke-attn-temp
```

**Step 4: Full training run**

```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --attn-temp-annealing --attn-temp-start 2.0 --attn-temp-end 1.0 \
  --wandb_group dm-innovation-track --name dm-attn-temp-annealing
```

**IMPORTANT:** Use the exact mandatory DrivAerML champion config above. Do NOT change any flags except the three new `--attn-temp-*` flags. The EMA=0.9995 + gc=0.5 pairing is critical for stability — do not modify either.

**Step 5: Report results in a PR comment:**
- W&B run ID and link
- Best `val_primary/surface_rel_l2_pct` vs baseline 3.833%
- Best epoch
- Observations on attention pattern behaviour over training

## Baseline

Current DrivAerML best:
- `val_primary/surface_rel_l2_pct`: **3.833%** (val) at epoch 511 — test **4.685%**
- Best PR: #3072 (eren — EMA=0.9995 + gc=0.5, 4L/512d/8H, AdamW lr=5e-4, T_max=30)
- External target: <3.71% (AB-UPT) — gap: 0.123 pp
- W&B run: ncl1dh88 (eren/dm-ema-9995-gc05)
- Reproduce:
  ```bash
  cd target/icml2026
  SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml \
    --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
    --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 \
    --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
    --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995
  ```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT vegeta:

## Results

### DrivAerML — Attention Temperature Annealing (τ: 2.0 → 1.0, linear schedule)

| Metric | Baseline (#3072) | This run | Delta |
|--------|------------------|----------|-------|
| `val_primary/surface_rel_l2_pct` | **3.833%** (ep 511) | **4.024%** (ep 485) | +0.19pp ❌ |
| `test_primary/surface_rel_l2_pct` | **4.685%** | **4.975%** | +0.29pp ❌ |
| External target (AB-UPT) | <3.71% | — | — |

- **W&B run:** [`ybh108ny`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ybh108ny)
- **Best epoch:** 485 (of 524 total, timeout-limited)
- **Tau at best epoch:** ~1.49 (annealing from 2.0 → 1.0 over 999 epochs)

### Command used
```bash
cd target/icml2026
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --attn-temp-annealing --attn-temp-start 2.0 --attn-temp-end 1.0 \
  --wandb-group dm-innovation-track --wandb-name "vegeta/dm-attn-temp-annealing" \
  --agent vegeta
```

### What happened

The attention temperature annealing feature **did not improve** over the baseline on the radford recipe. The best val metric was 4.024% vs the baseline 3.833% — a regression of ~0.19pp.

**Implementation details:**
- Temperature τ was applied to both the slice assignment softmax (divides logits by τ) and the QKV scaled dot-product attention (adjusts scale parameter) in every `TransolverAttention` layer
- Linear annealing from τ=2.0 (softer, broader attention) to τ=1.0 (standard sharpness) over the full training schedule
- The existing learnable per-head temperature parameter (initialized to 0.5) compounds with τ, so the effective initial temperature was ~1.0, annealing back to the default ~0.5

**Why it likely didn't work:**
1. **Learnable temperature already adapts.** The Transolver has a per-head learnable temperature parameter that can already discover the optimal softmax sharpness during training. Adding an external schedule may not provide additional benefit — or may fight the learned parameter.
2. **Slow annealing dilutes the effect.** With 999 epochs, τ only reached ~1.49 by the best epoch (485). The temperature at any given epoch barely differs from its predecessor, so the curriculum signal is very weak.
3. **QKV attention may not be load-bearing.** Research suggests that in Transolver-style architectures, the inter-slice QKV attention contributes minimally — the slice assignment mechanism is what matters. Applying τ to both may have introduced unnecessary perturbation.
4. **Interaction with cosine LR.** The linear temperature schedule is out of phase with the cosine LR restarts (T_max=30), potentially creating interference in the optimization landscape.

### Suggested follow-ups
- **Slice-only temperature annealing** — apply τ only to the slice assignment softmax, not the QKV attention. This targets the mechanism that matters most in Transolver.
- **Cosine temperature schedule** — align the temperature annealing with the cosine LR schedule for better synergy.
- **Shorter annealing window** — anneal τ from 2.0→1.0 over the first 60 epochs (2 cosine cycles) then hold at 1.0. This concentrates the curriculum where it matters most.
- **Freeze the learnable temperature** when external τ is active, for a cleaner experiment.

### Bug fix note
Fixed a pre-existing `UnboundLocalError` bug in `main()` where the local variable `primary_metric_key` (string) shadowed the module-level function `primary_metric_key()`, causing a crash on all DrivAerML runs. Renamed the local to `_primary_metric_str`.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3203 because ADVISOR: Closing — val=4.024% (+0.19pp / +5% vs 3.833%), test=4.975% (+0.29pp / +6.2%). The external annealed τ compounded with Transolver's existing learnable per-head temperature parameter, creating a convoluted optimization landscape rather than the intended soft→sharp progression. The -11% improvement seen on noam does not transfer to radford — likely due to noam lacking the learnable τ or having different optimizer dynamics. Attention temperature annealing added to DM dead ends. Bug fix for primary_metric_key shadowing noted — multiple students have now independently discovered this bug..
